rfd/0129-discovery-name-templating.md

```diff
@@ -0,0 +1,395 @@
+matching to identify resources.
+
+The same UX should apply to all `tsh` sub-commands that take a resource name
+argument. These commands shall support
+`tsh <sub-command> [name | prefix] [key1=value1,key2=value2,...]` syntax:
+
+- `tsh db login`
+- `tsh db connect`
+- `tsh db env`
+- `tsh db config`
+- `tsh proxy db`
+- `tsh proxy kube`
+- `tsh kube login`
+
+#### `tsh` examples
+
+To illustrate the new UX for `tsh` sub-commands, here is an example using
+`tsh db connect` to select a database (the same applies for other commands):
+
+```sh
+$ tsh db ls
+Name   Description         Allowed Users       Labels                      Connect 
+------ ------------------- ------------------- --------------------------- ------- 
+bar    RDS instance in ... [*] account-id=123456789012,region=us-west-1,env=dev,...
+bar    RDS instance in ... [*] account-id=123456789012,region=us-west-2,env=dev,...
+foo    RDS instance in ... [*] account-id=123456789012,region=us-west-1,env=prod,...
+
+# connect by prefix name
+$ tsh db connect --db-user=alice --db-name-postgres foo
+#...connects to "foo-rds-us-west-1-123456789012" by prefix...
+
+# ambiguous prefix name is an error
+$ tsh db connect --db-user=alice --db-name-postgres bar
+error: ambiguous database name could match multiple databases:
+Name                           Description               Protocol Type URI                                                   Allowed Users Labels                                                                                                                                    Connect 
+------------------------------ ------------------------- -------- ---- ----------------------------------------------------- ------------- ----------------------------------------------------------------------------------------------------------------------------------------- ------- 
+bar-rds-us-west-1-123456789012 RDS instance in us-west-1 postgres rds  bar.abcdefghijklmnop.us-west-1.rds.amazonaws.com:5432 [*]           account-id=123456789012,endpoint-type=instance,engine-version=13.10,engine=postgres,env=dev,region=us-west-1,teleport.dev/origin=dynamic          
+bar-rds-us-west-2-123456789012 RDS instance in us-west-2 postgres rds  bar.abcdefghijklmnop.us-west-2.rds.amazonaws.com:5432 [*]           account-id=123456789012,endpoint-type=instance,engine-version=13.10,engine=postgres,env=dev,region=us-west-2,teleport.dev/origin=dynamic          
+
+Hint: try addressing the database by its full name or by matching its labels (ex: tsh db connect key1=value1,key2=value2).
+Hint: use `tsh db ls -v` or `tsh db ls --format=[yaml | json]` to list all databases with verbose details.
+
+# resolve the error by connecting with an unambiguous prefix 
+$ tsh db connect --db-user=alice --db-name-postgres bar-rds-us-west-2
+#...connects to "bar-rds-us-west-2-123456789012" by prefix...
+
+# or connect by label(s)
+$ tsh db connect --db-user=alice --db-name-postgres region=us-west-2 
```

Comment:
Instead of "region=us-west-2" syntax, I think we should add proper support for `--query` flag that supports predicate language: https://goteleport.com/docs/reference/predicate-language/#resource-filtering. We already have it for some tsh commands, otherwise it will be inconsistent.

Reply:
I'll amend it. I have been working on implementing the tsh changes and was going to suggest this change anyway for simpler cli parsing.

Another thing I've come to realize is that if labels are used and the user has active certs from e.g. `tsh db login`, then we have to do an API call to determine if the active cert should be used, since certs don't contain the labels and the `<db>` cli arg might be a prefix for some other database that matches the labels

Reply:
c0e5a5bf2a7990029f8fa25c877e361ac765b373